app/logic/minor.py

```diff
@@ -68,34 +111,171 @@ def getProgramEngagementHistory(program_id, username, term_id):
+
+    courses = (Course.select(Course, courseMatchCase.alias("reqMatch"))
                      .join(CourseParticipant, on=(Course.id == CourseParticipant.course))
+                     .join(IndividualRequirement, JOIN.LEFT_OUTER, on=((IndividualRequirement.course == Course.id) & 
+                                                                       (IndividualRequirement.username == CourseParticipant.user) & 
+                                                                       (IndividualRequirement.term == Course.term)))
                      .where(CourseParticipant.user == username)
                      .group_by(Course.courseName, Course.term))
     
     # initialize default dict to store term descriptions as keys mapping to each
     # engagement's respective type, name, id, and term.
     terms = defaultdict(list)
     for course in courses:
-        terms[(course.term.description, course.term.id)].append({"name":course.courseName, "id":course.id, "type":"course", "term":course.term})
+        terms[(course.term.description, course.term.id)].append({
+                    "name":course.courseName, 
+                    "id":course.id, 
+                    "type":"course", 
+                    "matched": course.reqMatch,
+                    "term":course.term.id})
+
+    programMatchCase = Case(None, [(IndividualRequirement.program.is_null(True) , 0)], 1)
 
-    events = (Event.select(Event, Program)
+    events = (Event.select(Event, Program, programMatchCase.alias('reqMatch'))
                    .join(EventParticipant, on=(Event.id == EventParticipant.event)).switch()
                    .join(Program)
+                   .join(IndividualRequirement, JOIN.LEFT_OUTER, on=((IndividualRequirement.program == Program.id) &
+                                                                     (IndividualRequirement.username == EventParticipant.user) &
+                                                                     (IndividualRequirement.term == Event.term)))
                    .where(EventParticipant.user == username)
                    .group_by(Event.program, Event.term))
     
     for event in events:
-        terms[(event.term.description, event.term.id)].append({"name":event.program.programName, "id":event.program.id, "type":"program", "term":event.term})
-    
+        terms[(event.term.description, event.term.id)].append({"name":event.program.programName,
+                                                               "id":event.program.id,
+                                                               "type":"program",
+                                                               "matched": event.reqMatch,
+                                                               "term":event.term.id
+                                                              })
+
     # sorting the terms by the term id
     return dict(sorted(terms.items(), key=lambda x: x[0][1]))
```

Comment:
While using x for the lambda value is perhaps standard, I think we should rename the variable x to something more descriptive like `engagement` or something. Hard for people that aren't familiar with lambda functions to understand.